### PR TITLE
Add-on version handling

### PR DESCRIPTION
**Pull request by Nadahar**
Back in early 2025, I did a lot of work to try to made add-ons versioned, so that e.g. the marketplace would have a concept of the version of an add-on, allow upgrade, downgrade, update notification etc. This work stalled for a couple of reasons that I wasn't able to solve, and this is still the case, but I've kept the branch in case I see a solution to these issues some time in the future.

From time to time, the handling of add-on versions during upgrades, the fact that marketplace add-ons just "disappear" during upgrade, failures in the version range system and similar comes up, as it did again lately: https://community.openhab.org/t/marketplace-versioning-issues/169748

Knowing that I've solved many of those little things that are frustrations and obstacles for people to this day, a long time ago, I finally decided to try to extract just the version handling from the branch with the earlier work, and this PR is the result.

I've tried to remove everything that isn't directly related to the add-on version, but I've kept a few pure bugfixes just because, why not fix them when I know that they are there, and the fix has already been found..?

The changes in the PR centers about two new classes, `Version` and `VersionRange`. They make `BundleVersion` unnecessary, so it has been removed. `Version` and `VersionRange` closely follow the logic of OSGi's classes with the same names, but they are adjusted to fit with "OH's rules". In fact, I originally made them as extensions to the OSGi versions, but that caused so many problems during serialization/deserialization that I had to abandon that. They are still "compatible" with the OSGi classes and have methods for easy conversion both ways, without using string parsing.

`Version` is used in the `Addon` class itself, instead of basing everything on strings and string parsing. This makes it easy to compare, sort and "process" versions consistently throughout the system, and since it is easily convertible to/from OSGi's `Version`, it integrates well with version information obtained for bundles.

`VersonRange` is only used for the marketplace, but is also available system-wide, should the need arise elsewhere.

In addition, I've addressed a number of parsing issues that happen for some marketplace add-ons under some circumstances, every change I've made there is based on things I've "observed in the wild". The annotation and bugfix for `MarketplaceAddonHandler.supports()` is one of those, but I considered it important enough to include. When the NPE hits, the marketplace parsing is aborted and the user will see no marketplace add-ons at all.

Last, I included an unrelated fix, that I've long thought should have been brought in: Special treatment for GitHub links. Since GitHub is widely used for hosting marketplace resources, there have been numerous add-ons that have "used the wrong link" (I have notified those I've come across, so I think most of them have been rectified, but this will continue to happen in the future). GitHub have two ways to link a file, it's the "normal" link, which take you to a web page that shows the file, and then there's the "raw" link, which links only to the file itself. For the marketplace to work, people must use the "raw" link, but it's easy to miss. The fix here inspects links that are GitHub only, and checks whether they are "raw" or not. If they aren't, it adds a query parameter that GitHub provides that makes sure that when the link is used, you're forwarded to the "raw" file instead of getting the webpage view of the file.

### CODE REVIEW

**Comment by Nadahar** [comment]
Since I just discovered [the problem with marketplace link resolution](https://community.openhab.org/t/marketplace-link-resolution/169828), I added the fix to this PR. It fits nicely to the other fixes, and concerns the same code.

**Comment by Nadahar** [comment]
I've rebased this PR and resolved where this and #5710 "stepped on each other's toes".

**Comment by mherwege** [comment]
I am just wondering if there is anything specific to our version definitions and ranges that is not covered by standard maven versions and ranges. I put some version comparison logic in a recent UpgradeTool PR where I directly use the maven classes: https://github.com/openhab/openhab-core/pull/5379. Here is a reference to the maven classes: https://maven.apache.org/ref/3.6.2/maven-artifact/apidocs/org/apache/maven/artifact/versioning/package-summary.html

**Comment by Nadahar** [comment]
> I am just wondering if there is anything specific to our version definitions and ranges that is not covered by standard maven versions and ranges.

I wrote this code 1.5 years ago, so the details aren't exactly fresh, but yes, I believe there are some differences. One is that this code takes into account the "special OH rules" for the order of milestone -> release candidate -> release, while standard Maven just do alphabetical comparison or something like that. Another is that I dislike Maven's shorthand form, where if you only specify one version, that automatically means "this or newer". I think that's ambiguous, it would just as well mean "only this version specifically". The current logic doesn't accept anything that doesn't have two endpoints, so it would just consider it invalid. In this PR however, I implemented support for open-ended, but you'd have to indicate that it's open-ended: `[5.2.0;]` etc. 

This was my choice, but I think it's needed, because you often have no idea when it will stop working., which makes people guess, which again leads to marketplace add-ons just suddenly "disappearing" when people upgrade for no other reason than that nobody has bumped the upper limit. When you don't know, I think it's better to make it open-ended, and the day it breaks on some version, it's easy to modify the add-on entry and put the appropriate upper endpoint there.

Maven also accepts a "range" without any brackets/parentheses, as I mentioned above, which "means" this or any newer version. That sort of works for Maven because it's specified in dedicated "version" fields, however, we need to "recognize it" as part of a larger text (the title), so allowing such expressions would be very unfortunate. By having a stricter syntax, we can be reasonably certain that when we find a "match" in the title, it *is* actually a version range, and not something different.

That said, I've made `Version` capable of parsing strings as well, so if you wish to parse just a single version and not a range, you can just use `Version.valueOf()`.

My idea when I made this was to have one definition that is used everywhere, avoiding confusion and inconsistency.

edit: Don't forget that we're not only dealing with Maven. This integrates "seamlessly" with OSGi's version and version range concepts, and I must admit that I had more focus on aligning with OSGi than with Maven.

File: bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/Addon.java [hunk 1]
```diff
@@ -436,7 +442,11 @@ public Builder withKeywords(String keywords) {
...
 
-        public Builder withCountries(List<String> countries) {
+        public @Nullable List<@NonNull String> getCountries() {
+            return countries;
+        }
+
+        public Builder withCountries(List<@NonNull String> countries) {
             this.countries = countries;
             return this;
         }
```
[the remark this answers is not in the document]
Comment (Nadahar) [reply bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/Addon.java]: This class doesn't use `@NonNullByDefault`, so there's no implicit non-null requirement here. All builder methods support `null` generally, because that's how you "unset" a parameter. Correcting this would mean annotating all of them, which should be done if `@NonNullByDefault` is applied. I was tempted to do it, but tried to restrain myself from making unrelated (to this PR) changes.

Looking at the other methods, there is a lot of inconsistency, some are annotated, some aren't, so I guess I won't "hurt" to add annotation here as well. But, it's not "wrong" as it is.

File: bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/AbstractRemoteAddonService.java [hunk 2]
```diff
@@ -67,21 +69,11 @@ public abstract class AbstractRemoteAddonService implements AddonService {
...
-            BundleVersion version2 = new BundleVersion(addon2.getVersion().replaceAll("-.*", ".0"));
-
-            // prefer newer version over older
-            return version2.compareTo(version1);
-        } catch (IllegalArgumentException e) {
-            // assume they are equal (for ordering) if we can't compare the versions
-            return 0;
-        }
+        Version v2 = addon2.getVersion();
+        return v2 == null ? 1 : v2.compareTo(addon1.getVersion());
```
[the remark this answers is not in the document]
Comment (Nadahar) [reply bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/AbstractRemoteAddonService.java]: Excellent point! Fixed.

File: bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/Addon.java [hunk 3]
```diff
@@ -461,26 +471,34 @@ public Builder withImageLink(@Nullable String imageLink) {
...
 
-        public Builder withProperty(String key, Object value) {
-            this.properties.put(key, value);
+        public Builder withProperty(@NonNull String key, @NonNull Object value) {
+            Map<@NonNull String, @NonNull Object> props = this.properties;
+            if (props == null) {
+                props = new HashMap<>();
+            }
+            props.put(key, value);
+            this.properties = props;
```
[the remark this answers is not in the document]
Comment (Nadahar) [reply bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/Addon.java:480]: Fixed.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/VersionRange.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        int h = hash;
+        if (h != 0) {
+            return h;
+        }
+        return h = Objects.hash(leftInclusive, left, right, rightInclusive);
+    }
```
[the remark this answers is not in the document]
Comment (Nadahar) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/VersionRange.java:203]: Duh! I got the same thing correct in `Version` though, don't know how this happened. Fixed.

**Comment by mherwege** [comment]
> One is that this code takes into account the "special OH rules" for the order of milestone -> release candidate -> release, while standard Maven just do alphabetical comparison or something like that.

I think Maven does that right. It is OSGI that cannot handle this properly. OSGI versions are `major.minor.micro.qualifier` where the `qualifier` is a string and sorted alphabetically. OSGI does not have any notion of snapshots, milestones, release candidates or anything. The version ranges as we define them now have 4 digits in the marketplace. OSGI only allows 3.

> Another is that I dislike Maven's shorthand form, where if you only specify one version, that automatically means "this or newer". I think that's ambiguous, it would just as well mean "only this version specifically".

We can like or dislike this, but it is a fairly standard way of doing things. It also allows the full syntax I believe. OSGI, depending on the tool used (BND, Eclipse) may actually do the same thing.

I kind of feel like we are reinventing the wheel. Is there anything OSGI specific in your code?

**Comment by Nadahar** [comment]
> I think Maven does that right. It is OSGI that cannot handle this properly. OSGI versions are `major.minor.micro.qualifier` where the `qualifier` is a string and sorted alphabetically. OSGI does not have any notion of snapshots, milestones, release candidates or anything. The version ranges as we define them now have 4 digits in the marketplace. OSGI only allows 3.

Maven don't have these concepts either. What Maven has, is the very messy `SNAPSHOT` functionality, because it doesn't point to any particular version. I see that it's practical, but it also causes a lot of headaches, and I understand why OSGi doesn't support it. Not much outside of Maven has a concept like this.

I've never heard anything about Maven having rules for milestones, release candidates etc... I think it only treats whatever you put as the "qualifier" as an alphanumeric value, except the magical `SNAPSHOT` version.

Regardless, Maven has this other "huge problem" that it doesn't differentiate between a version and a version range. "Everything" in a way is a version range. I think that's very messy.

> I kind of feel like we are reinventing the wheel. Is there anything OSGI specific in your code?

I don't know which wheel we're reinventing, or what you mean by "OSGi specific", I originally made these as extension of OSGi's `Version` and `VersionRange`, so they share the "fundamental structure". The reason for extending them was to implement OH specific rules for comparison/order and parsing. I then got into a lot of trouble when deserializing, because the "wrong types" would be created, so I made them independent but "compatible" with to/from conversions, since they have the same internal structure.

Today we have `BundleVersion` that fill this role for the marketplace, and in the rest of the system, it's all string handling so that the exact "rules" depends on what however wrote that specific code chose.

I assume that what you want to use instead is one of these? 

https://maven.apache.org/ref/3.9.16/maven-artifact/xref/org/apache/maven/artifact/versioning/DefaultArtifactVersion.html
https://maven.apache.org/ref/3.9.16/maven-artifact/xref/org/apache/maven/artifact/versioning/ComparableVersion.html

I must admit that I haven't studied in details how they work, because they are, in my view, quite mismatched to what is needed here.

It should also be said that I have done a lot more around versioning and the marketplace in a never completed branch (because I met obstacles I don't know how to get around), where both `Version` and `VersionRange` play significant roles and is part of the foundation the whole thing is built on. So, while this might seem a bit overkill when isolated like this, it wasn't when I wrote them.

**Comment by Nadahar** [comment]
> The version ranges as we define them now have 4 digits in the marketplace. OSGI only allows 3.

This is a misunderstanding. The last "number" is just the alphanumeric qualifier. The reason it is used on the marketplace is that it's the only way to get `BundleVersion` to treat the boundaries like we "want it to". It's just a workaround. With this implementation, we could use how few numbers we'd like. `5` is interpreted as `5.0.0`, `5.2` as `5.2.0` etc.

File: bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/dto/AddonDTO.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        this.verifiedAuthor = addon.isVerifiedAuthor();
+        this.installed = addon.isInstalled();
+        this.type = addon.getType();
+        this.description = addon.getDescription();
+        this.detailedDescription = addon.getDetailedDescription();
+        this.configDescriptionURI = addon.getConfigDescriptionURI();
+        this.keywords = addon.getKeywords();
+        List<@NonNull String> stringList = addon.getCountries();
+        if (!stringList.isEmpty()) {
+            this.countries = stringList;
```
Comment (wborn) [inline bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/dto/AddonDTO.java:86]: Could we preserve the existing REST response contract by always using empty collections instead of leaving these fields `null`?

MainUI currently assumes that fields such as `loggerPackages` and `properties` are present and accesses them using expressions such as `addon.loggerPackages.length` and `Object.keys(addon.properties).length`.

Because Gson normally omits `null` fields, add-ons without these values may no longer contain `loggerPackages`, `properties`, or `countries` in the response, potentially causing errors on the add-on details page.

Initializing these fields with empty collections, or assigning them unconditionally, would preserve the previous response shape.

File: bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/AbstractRemoteAddonService.java [hunk 6]
```diff
@@ -67,21 +69,12 @@ public abstract class AbstractRemoteAddonService implements AddonService {
...
-
-            // prefer newer version over older
-            return version2.compareTo(version1);
-        } catch (IllegalArgumentException e) {
-            // assume they are equal (for ordering) if we can't compare the versions
-            return 0;
-        }
+        Version v1 = addon1.getVersion();
+        Version v2 = addon2.getVersion();
+        return v1 == null && v2 == null ? 0 : v2 == null ? 1 : v2.compareTo(v1);
```
Comment (wborn) [inline bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/AbstractRemoteAddonService.java:74]: This comparator appears to violate the `Comparator` contract when exactly one add-on has no version.

For a versioned add-on `a` and an unversioned add-on `b`, both of these comparisons return a positive value:

```java
compare(a, b)
compare(b, a)
```

The first result comes from `v2 == null`, while the second comes from `Version.compareTo(null)` returning `1`. This makes the comparator non-antisymmetric and may result in incorrect ordering or an `IllegalArgumentException` during sorting.

Could both one-null cases be handled explicitly? It would also be useful to add tests for `(null, null)`, `(null, version)`, and `(version, null)`.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/Version.java [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+        if (ql >= 0 && oql >= 0) {
+            // Both are positive integers, compare numerically
+            return Long.compare(ql, oql);
+        }
+
+        boolean ss = SNAPSHOT_PATTERN.matcher(qualifier).matches();
+        boolean oss = SNAPSHOT_PATTERN.matcher(other.qualifier).matches();
+        if (ql < 0 && oql < 0) {
+            // Both aren't positive integers, snapshots are newer, otherwise do a simple string comparison
+            if (ss) {
```
Comment (wborn) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/common/Version.java:319]: The snapshot comparison appears to be non-antisymmetric when both qualifiers are snapshot variants with different capitalization.

In that case, both `ss` and `oss` are `true`, so the first `if (ss)` returns `1`. Reversing the operands also returns `1`. The later branch intended to compare two differently capitalized snapshot qualifiers is therefore unreachable.

Could the `ss && oss` case be handled before the cases where only one qualifier is a snapshot? A test that verifies comparator symmetry for `SNAPSHOT` and `snapshot` would help prevent this regression.

File: bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/internal/json/JsonAddonService.java [hunk 8]
```diff
@@ -199,16 +201,18 @@ private Addon fromAddonEntry(AddonEntryDTO addonEntry) {
...
         boolean compatible = true;
         try {
-            compatible = coreVersion.inRange(addonEntry.compatibleVersions);
+            compatible = VersionRange.valueOf(addonEntry.compatibleVersions).includes(coreVersion);
         } catch (IllegalArgumentException e) {
             logger.debug("Failed to determine compatibility for addon {}: {}", addonEntry.id, e.getMessage());
         }
 
+        Version v = addonEntry.version == null || addonEntry.version.isBlank() ? null
+                : Version.valueOf(addonEntry.version);
```
Comment (wborn) [inline bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/internal/json/JsonAddonService.java:210]: Could version parsing be isolated per repository entry?

Previously, the version was accepted as an arbitrary string. With this change, a single entry containing a version syntax unsupported by `Version.valueOf()` can throw from inside the stream mapping and abort processing of the entire repository.

A malformed version should probably be logged and treated as unknown, or cause only that individual entry to be skipped. It should not make all add-ons from the source disappear.

File: bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/Addon.java [hunk 9]
```diff
@@ -461,26 +471,34 @@ public Builder withImageLink(@Nullable String imageLink) {
...
             return this;
         }
 
-        public Builder withProperties(Map<String, Object> properties) {
-            this.properties.putAll(properties);
+        public Builder withProperties(@Nullable Map<@NonNull String, @NonNull Object> properties) {
+            this.properties = properties;
```
Comment (wborn) [inline bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/Addon.java:485]: `withProperties()` previously copied the supplied values into the builder’s mutable map. It now stores the supplied map directly.

Apart from changing the previous merge semantics, this can also cause an unexpected exception:

```java
builder.withProperties(Map.of("one", 1))
       .withProperty("two", 2);
```

Since `Map.of()` is immutable, `withProperty()` will throw `UnsupportedOperationException`. A mutable map supplied by a caller can also be modified indirectly through the builder.

Could the supplied map be defensively copied into a new `LinkedHashMap`? A test using an immutable map followed by `withProperty()` would cover this case.

**Review by wborn — COMMENTED** [review]
**Disclosure:** This review was generated with AI assistance. The findings and comments were reviewed before submission.

File: bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/Addon.java (hunk 9 above)
Comment (Nadahar) [reply bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/Addon.java:485]: Sure, we can make defensive copies all we want. My assumption was that people will either use `withProperties()` or `withProperty()`, not both, and if they must use both, then they'd better not provide an immutable map first. But I can make it handle "anything we want". When it comes to the merge semantics, the primary reason for setting the map directly, is that it should be possible to "null" the properties. Previously, you couldn't do that, you could only add. But that can be solved too, by have two different paths, where `null` means "set it to null", while non-null means "add to what's already here".

File: bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/dto/AddonDTO.java (hunk 5 above)
Comment (Nadahar) [reply bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/dto/AddonDTO.java:86]: It was intended for them to be missing from the DTO if empty, like most other fields. I've already made a PR in MainUI that makes it handle that: https://github.com/openhab/openhab-webui/pull/4365. 

It's very "arbitrary" when some fields have to be initialized if empty, while most don't, which is why I preferred to unity the behavior. Also, since there never was a DTO class in the past, the serialization wasn't exactly "designed", but became what it became based on hot `Addon` itself behaved.

**Comment by Nadahar** [comment]
@wborn @mherwege I'm a bit unsure of what to do here. I could address the above comments, but if @mherwege is opposed to the whole idea of what I'm doing, is it pointless?

I don't mind addressing the points, but I don't think it's "that much fun" that I'll gladly do it knowing that it will never go anywhere.

As said above, I have a lot more work that is built upon the above, so some of the "details" that might seem excessive are there because they are needed in the code that is built on top of it. That doesn't make them "bad" in themselves as I see it, but seeing the rest of the code might make the motivation a bit clearer. I won't ask you to try to review that though, both since it's quite complex, and because it's not finished and has some unresolved challenges in it. But those challenges aren't about the version handling itself, but about how to store the version information in a way that it won't be lost (various add-on types are stored in very different ways locally, and not all of them have an obvious way to store "extra information"), and it's about `FeatureInstaller` and basically all the hard-coded behavior that exist when it comes to KARs and add-ons. Except for those two "areas", my "versioning" branch works, it does allow upgrading or downgrading add-ons etc.

**Comment by mherwege** [comment]
> I could address the above comments, but if @mherwege is opposed to the whole idea of what I'm doing, is it pointless?

It is not that I am opposed to what you are doing. I questionned if creating specific logic to be maintained going forward is the right thing to do, if there may be other ways to achieve this. You seem to think there is none. I must say I am not yet entirely convinced about that, but that does not mean I will oppose this. Overall, the code looks good to me.

**Comment by Nadahar** [comment]
> You seem to think there is none.

There is none that I've been able to identify that fits the current requirements/needs. I don't think that logic will require much maintenance until there is a change in OHs "versioning scheme", and if that happens, I would think that it's beneficial to be able to adjust the logic accordingly, in one place, and then carry on.
